polyfill/lib/ecmascript.mjs

```diff
@@ -4017,6 +3751,708 @@ export function DifferenceZonedDateTime(ns1, ns2, timeZoneRec, calendarRec, larg
+    if (didExpandToEnd) {
+      duration = endDuration;
+    } else {
+      // NOT at end-of-unit. Stop looking for bubbling
+      break;
+    }
+  }
+
+  return duration;
+}
+
+function RoundRelativeDuration(
+  duration,
+  destEpochNs,
+  dateTime,
+  calendarRec,
+  timeZoneRec,
+  largestUnit,
+  increment,
+  smallestUnit,
+  roundingMode
+) {
+  // The duration must already be balanced. This should be achieved by calling
+  // one of the non-rounding since/until internal methods prior. It's okay to
+  // have a bottom-heavy weeks because weeks don't bubble-up into months. It's
+  // okay to have >24 hour day assuming the final day of relativeTo+duration has
+  // >24 hours in its timezone. (should automatically end up like this if using
+  // non-rounding since/until internal methods prior)
+  const irregularLengthUnit = IsCalendarUnit(smallestUnit) || (timeZoneRec && smallestUnit === 'day');
+  const sign =
+    DurationSign(duration.years, duration.months, duration.weeks, duration.days, duration.norm.sign()) < 0 ? -1 : 1;
+
+  let nudgeResult;
+  if (irregularLengthUnit) {
+    // Rounding an irregular-length unit? Use epoch-nanosecond-bounding technique
+    nudgeResult = NudgeToCalendarUnit(
+      sign,
+      duration,
+      destEpochNs,
+      dateTime,
+      calendarRec,
+      timeZoneRec,
+      increment,
+      smallestUnit,
+      roundingMode
+    );
+  } else if (timeZoneRec) {
+    // Special-case for rounding time units within a zoned day. total() never
+    // takes this path because largestUnit is then also a time unit, so
+    // DifferenceZonedDateTimeWithRounding uses Instant math
+    nudgeResult = NudgeToZonedTime(
+      sign,
+      duration,
+      dateTime,
+      calendarRec,
+      timeZoneRec,
+      increment,
+      smallestUnit,
+      roundingMode
+    );
+  } else {
+    // Rounding uniform-length days/hours/minutes/etc units. Simple nanosecond
+    // math. years/months/weeks unchanged
+    nudgeResult = NudgeToDayOrTime(duration, destEpochNs, increment, smallestUnit, roundingMode);
+  }
+
+  duration = nudgeResult.duration;
+  // Did nudging cause the duration to expand to the next day or larger?
+  // Bubble-up smaller calendar units into higher ones, except for weeks, which
+  // don't balance up into months
+  if (nudgeResult.didExpandCalendarUnit && smallestUnit !== 'week') {
+    duration = BubbleRelativeDuration(
+      sign,
+      duration,
+      nudgeResult.nudgedEpochNs, // The destEpochNs after expanding/contracting
+      dateTime,
+      calendarRec,
+      timeZoneRec,
+      largestUnit, // where to STOP bubbling
+      LargerOfTwoTemporalUnits(smallestUnit, 'day') // where to START bubbling-up from
+    );
+  }
+
+  const {
+    days: deltaDays,
+    hours,
+    minutes,
+    seconds,
+    milliseconds,
+    microseconds,
+    nanoseconds
+  } = BalanceTimeDuration(duration.norm, largestUnit);
+  const days = duration.days + deltaDays;
```

Comment:
I think pushing smushing deltaDays into the final result at this late stage could result in an unbalanced duration (because if days overflows months, which overflow years...). I think we are ONLY interested in time values at this point. So, we should call BalanceTimeDuration with a largestUnit that is never larger than hours. If `norm` comes out to 24:30, that's fine, that means relativeTo was a ZonedDateTime and the relativeTo + initial duration landed in a 25 hour day.

Reply:
I did play around with this while translating the pseudocode into this code. It ended up having to be different than you did it in the Fullcalendar polyfill because of using normalized duration math in NudgeToDayOrTime; without this, we'd otherwise have to pass largestUnit into NudgeToDayOrTime and split out the days if largestUnit was days or larger.

That's possible but more complicated (I'll paste the code diff below). It doesn't affect any test262 tests, so if you're sure this is more correct then we should be able to come up with a test case where it makes a difference.

Reply:
Diff:
```diff
diff --git a/polyfill/lib/ecmascript.mjs b/polyfill/lib/ecmascript.mjs
index a7c8357f..6ae33bc3 100644
--- a/polyfill/lib/ecmascript.mjs
+++ b/polyfill/lib/ecmascript.mjs
@@ -4032,24 +4032,31 @@ function NudgeToZonedTime(sign, duration, dateTime, calendarRec, timeZoneRec, in
 }
 
 // Converts all fields to nanoseconds and does integer rounding.
-function NudgeToDayOrTime(duration, destEpochNs, increment, unit, roundingMode) {
+function NudgeToDayOrTime(duration, destEpochNs, largestUnit, increment, smallestUnit, roundingMode) {
   // unit must be day or smaller
 
   const norm = duration.norm.add24HourDays(duration.days);
   // Convert to nanoseconds and round
-  const unitLength = NS_PER_TIME_UNIT.get(unit);
+  const unitLength = NS_PER_TIME_UNIT.get(smallestUnit);
   const total = norm.fdiv(unitLength);
   const roundedNorm = norm.round(increment * unitLength, roundingMode);
   const diffNorm = roundedNorm.subtract(norm);
 
   // Determine if whole days expanded
   const { quotient: wholeDays } = norm.divmod(DAY_NANOS);
-  const { quotient: roundedWholeDays } = roundedNorm.divmod(DAY_NANOS);
+  const { quotient: roundedWholeDays, remainder: roundedDaysRemainder } = roundedNorm.divmod(DAY_NANOS);
   const didExpandDays = MathSign(roundedWholeDays - wholeDays) === norm.sign();
 
   const nudgedEpochNs = diffNorm.addToEpochNs(destEpochNs);
 
-  duration = { ...duration, days: 0, norm: roundedNorm };
+  let days = 0;
+  let remainder = roundedNorm;
+  if (LargerOfTwoTemporalUnits(largestUnit, 'day') === largestUnit) {
+    days = roundedWholeDays;
+    remainder = roundedDaysRemainder;
+  }
+
+  duration = { ...duration, days, norm: remainder };
   return {
     duration,
     total,
@@ -4232,7 +4239,7 @@ function RoundRelativeDuration(
   } else {
     // Rounding uniform-length days/hours/minutes/etc units. Simple nanosecond
     // math. years/months/weeks unchanged
-    nudgeResult = NudgeToDayOrTime(duration, destEpochNs, increment, smallestUnit, roundingMode);
+    nudgeResult = NudgeToDayOrTime(duration, destEpochNs, largestUnit, increment, smallestUnit, roundingMode);
   }
 
   duration = nudgeResult.duration;
@@ -4252,21 +4259,18 @@ function RoundRelativeDuration(
     );
   }
 
-  const {
-    days: deltaDays,
-    hours,
-    minutes,
-    seconds,
-    milliseconds,
-    microseconds,
-    nanoseconds
-  } = BalanceTimeDuration(duration.norm, largestUnit);
-  const days = duration.days + deltaDays;
+  if (IsCalendarUnit(largestUnit) || largestUnit === 'day') {
+    largestUnit = 'hour';
+  }
+  const { hours, minutes, seconds, milliseconds, microseconds, nanoseconds } = BalanceTimeDuration(
+    duration.norm,
+    largestUnit
+  );
   return {
     years: duration.years,
     months: duration.months,
     weeks: duration.weeks,
-    days,
+    days: duration.days,
     hours,
     minutes,
     seconds,
```

Reply:
Tentatively, I think they are the same. BubbleRelativeDuration always returns a normalized time component of 0. BubbleRelativeDuration is skipped if smallestUnit is week (then the normalized time component is also 0) or didExpandCalendarUnit is false. So the only way to have a nonzero normalized time at the BalanceTimeDuration call is if didExpandCalendarUnit is false, or if we went through NudgeToDayOrTime and expanded hours into days but not into weeks. In neither case do we get an unbalanced duration.

(Although I agree this is a bit convoluted. Maybe the more complicated code is actually easier to reason about.)

Reply:
I might have found a test case where this could happen, but _if_ this is a bug, it is currently inside NudgeToZonedTime and not here.
```js
const duration = Temporal.Duration.from({ hours: 24, nanoseconds: 5 });
const relativeTo = Temporal.ZonedDateTime.from('2000-10-29T00[America/Vancouver]');
// note: relativeTo.hoursInDay === 25
const result = duration.round({
  largestUnit: 'days',
  smallestUnit: 'minutes',
  roundingMode: 'expand',
  roundingIncrement: 30,
  relativeTo
});
console.log(result.toString());
```
This gives "P1DT30M" (and so does the Fullcalendar polyfill) but I'd tentatively expect it to give "PT24H30M"...

Reply:
PT24H30M was the result that we'd get from this code without the new optimization, so this seems like it's a regression and should have test coverage.

Reply:
Hi @ptomato, I looked into this and there are number of issues at work:
- The fullcalendar polyfill is giving wrong results because of an unrelated bug. It was a bug I identified while devising the formal algorithm but did not backport the fix to the polyfill itself. [Right here if you're curious](https://github.com/fullcalendar/temporal-polyfill/blob/main/packages/temporal-polyfill/src/internal/round.ts#L588). The algorithm in the gist does not have this bug.
- Separately, I identified a bug in this polyfill's implementation of`NudgeToZonedTime`. I've written it as an inline comment (`TimeDuration.ZERO` should be used instead of `duration.norm`).
- Once that bug is fixed, the bug *persists* because of a second issue, the issue with `deltaDays` and `largestUnit` at the end of `RoundRelativeDuration`.  I applied the diff you wrote to a checkout of proposal-temporal (including having `NudgeToDayOrTime` accept largestUnit). I also applied the `TimeDuration.ZERO` fix. These two fixes together fix the "PT24H30M" bug while ensuring all test262 tests passing.

Reply:
@arshaw Thanks so much for figuring that out. You really saved me a bunch of work!

I'll apply both of those fixes and write up the necessary changes in the spec text. I've already added the test case to https://github.com/tc39/test262/pull/4041.